src/hooks/useTodos.ts

```
@@ -0,0 +1,187 @@
+import { useCallback, useEffect, useMemo, useState } from 'react';
+import { filterTodos } from '../utils/services';
+import * as postService from '../api/todos';
+
+import { Filters, Todo } from '../types/Todo';
+
+export const useTodos = () => {
+  const [todos, setTodos] = useState<Todo[]>([]);
+  const [tempTodo, setTempTodo] = useState<Todo | null>(null);
+  const [isLoading, setIsLoading] = useState(false);
+  const [errorMessage, setErrorMessage] = useState<string | null>(null);
+  const [activeFilter, setActiveFilter] = useState<Filters>(Filters.All);
+  const [editingTodos, setEditingTodos] = useState<number[]>([]);
+  const [renamingTodo, setRenamingTodo] = useState<number | null>(null);
+
+  const isAllCompleted = todos.every(todo => todo.completed);
+
+  const filtered = useMemo(
+    () => filterTodos(todos, activeFilter),
+    [todos, activeFilter],
+  );
+
+  const addTodo = useCallback(async (title: string) => {
+    setTempTodo({
+      id: 0,
+      userId: postService.USER_ID,
+      title: title,
+      completed: false,
+    });
+    setIsLoading(true);
+
+    try {
+      try {
+        const newTodo = await postService.createTodo(title);
+
+        setEditingTodos(current => [...current, 0]);
+
+        setTodos(currentTodos => [...currentTodos, newTodo]);
+      } catch {
+        setErrorMessage('Unable to add a todo');
+        throw new Error('Unable to add a todo');
```

Do we really want to throw an error here? If not caught by error boundary, this would break the app.
\+ the same question about similar code in `updateTodo`.